cdc/kv/region_worker.go

```diff
@@ -0,0 +1,390 @@
+				logPanic("The CommitTs must be greater than the resolvedTs",
+					zap.String("Event Type", "COMMITTED"),
+					zap.Uint64("CommitTs", entry.CommitTs),
+					zap.Uint64("resolvedTs", state.lastResolvedTs),
+					zap.Uint64("regionID", regionID))
+				return errUnreachable
+			}
+			select {
+			case w.outputCh <- revent:
+				metricSendEventCommittedCounter.Inc()
+			case <-ctx.Done():
+				return errors.Trace(ctx.Err())
+			}
+		case cdcpb.Event_PREWRITE:
+			metricPullEventPrewriteCounter.Inc()
+			state.matcher.putPrewriteRow(entry)
+		case cdcpb.Event_COMMIT:
+			metricPullEventCommitCounter.Inc()
+			if entry.CommitTs <= state.lastResolvedTs {
+				logPanic("The CommitTs must be greater than the resolvedTs",
+					zap.String("Event Type", "COMMIT"),
+					zap.Uint64("CommitTs", entry.CommitTs),
+					zap.Uint64("resolvedTs", state.lastResolvedTs),
+					zap.Uint64("regionID", regionID))
+				return errUnreachable
+			}
+			ok := state.matcher.matchRow(entry)
+			if !ok {
+				if !state.initialized {
+					state.matcher.cacheCommitRow(entry)
+					continue
+				}
+				return cerror.ErrPrewriteNotMatch.GenWithStackByArgs(entry.GetKey(), entry.GetStartTs())
+			}
+
+			revent, err := assembleRowEvent(regionID, entry, w.enableOldValue)
+			if err != nil {
+				return errors.Trace(err)
+			}
+
+			select {
+			case w.outputCh <- revent:
+				metricSendEventCommitCounter.Inc()
+			case <-ctx.Done():
+				return errors.Trace(ctx.Err())
+			}
+		case cdcpb.Event_ROLLBACK:
+			metricPullEventRollbackCounter.Inc()
+			state.matcher.rollbackRow(entry)
+		}
+	}
+	return nil
+}
+
+func (w *regionWorker) handleResolvedTs(
+	ctx context.Context,
+	resolvedTs uint64,
+	state *regionFeedState,
+	metricSendEventResolvedCounter prometheus.Counter,
+) error {
+	if !state.initialized {
+		return nil
+	}
+	regionID := state.sri.verID.GetID()
+	if resolvedTs < state.lastResolvedTs {
+		log.Warn("The resolvedTs is fallen back in kvclient",
+			zap.String("Event Type", "RESOLVED"),
+			zap.Uint64("resolvedTs", resolvedTs),
+			zap.Uint64("lastResolvedTs", state.lastResolvedTs),
+			zap.Uint64("regionID", regionID))
+		return nil
+	}
+	// emit a checkpointTs
+	revent := &model.RegionFeedEvent{
+		RegionID: regionID,
+		Resolved: &model.ResolvedSpan{
+			Span:       state.sri.span,
+			ResolvedTs: resolvedTs,
+		},
+	}
+	state.lastResolvedTs = resolvedTs
+
+	select {
+	case w.outputCh <- revent:
+		metricSendEventResolvedCounter.Inc()
+	case <-ctx.Done():
+		return errors.Trace(ctx.Err())
+	}
+	return nil
+}
+
+// evictAllRegions is used when gRPC stream meets error and re-establish, notify
+// all existing regions to re-establish
+func (w *regionWorker) evictAllRegions(ctx context.Context) error {
+	w.statesLock.Lock()
+	defer w.statesLock.Unlock()
+	for _, state := range w.regionStates {
+		log.Info("on region fail fired", zap.Reflect("state", state))
+		err := w.session.onRegionFail(ctx, regionErrorInfo{
+			singleRegionInfo: state.sri,
```

Comment:
Should we lock state before reading its field?

Reply:
addressed